### PR TITLE
fix: specialization

### LINKED ISSUE #1402: specialization not fully supported
cc https://rust-lang.github.io/rfcs/1210-impl-specialization.html#the-default-keyword
```rust
default fn 
default type
```

### PR DESCRIPTION
**Pull request by csmoe**
Closes #1402 
r? @matklad

### CODE REVIEW

File: crates/ra_syntax/tests/data/parser/inline/ok/0132_default_type.rs [hunk 1]
```diff
@@ -0,0 +1 @@
+default type T = Foo;
```
Comment (bjorn3) [inline crates/ra_syntax/tests/data/parser/inline/ok/0132_default_type.rs]: Rustc rejects this: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=6b9352dc533fdd2ab96f3b3e6f0a38f3

Comment (csmoe) [reply crates/ra_syntax/tests/data/parser/inline/ok/0132_default_type.rs]: `default type` must be placed inside `impl` as:
```rust
impl Trait for Foo {
    default T = Debug;
}
```
cc https://rust-lang.github.io/rfcs/1210-impl-specialization.html#the-default-keyword

Comment (bjorn3) [reply crates/ra_syntax/tests/data/parser/inline/ok/0132_default_type.rs]: I know, but here it isnt, yet ra seems to parse it fine.

Comment (csmoe) [reply crates/ra_syntax/tests/data/parser/inline/ok/0132_default_type.rs]: oops, I think I misunderstood something, fix soon.

**Comment by matklad** [comment]
changes look good to me, but CI's failing

**Comment by csmoe** [comment]
@matklad CI was unhappy with `test_err default fn foo() {}`. 
I found that `ItemFlavor` is not concrete enough in this case, as the original code mark `impl {}` as `Mod`, but `default fn foo() {}` isn't allowed in `Mod`, neither `Trait`.
so I'm imaging two approach:
1. add a `in_impl: bool`
2. add a variant `ItemFlavor::Impl`
which one do you prefer? or better alternatives?

**Comment by matklad** [comment]
I'd say we should accept default in all cases, and provided a targeted error message and a fix in validation layer

**Comment by csmoe** [comment]
oops, `default type` failed too.
so @bjorn3 , I'd let them be parsed in `mod {}` here as @matklad said at first.

**Pull request merged**